lms/djangoapps/courseware/views/views.py

```diff
@@ -749,42 +739,101 @@ def _progress(request, course_key, student_id):
         'staff_access': staff_access,
         'student': student,
         'passed': is_course_passed(course, grade_summary),
-        'show_generate_cert_btn': show_generate_cert_btn,
         'credit_course_requirements': _credit_course_requirements(course_key, student),
-        'missing_required_verification': missing_required_verification,
-        'certificate_invalidated': False,
-        'enrollment_mode': enrollment_mode,
+        'certificate_data': _get_cert_data(student, course, course_key, is_active, enrollment_mode)
     }
 
-    if show_generate_cert_btn:
-        # If current certificate is invalidated by instructor
-        # then show the certificate invalidated message.
-        context.update({
-            'certificate_invalidated': certs_api.is_certificate_invalid(student, course_key)
-        })
-
-        cert_status = certs_api.certificate_downloadable_status(student, course_key)
-        context.update(cert_status)
-        # showing the certificate web view button if feature flags are enabled.
-        if certs_api.has_html_certificates_enabled(course_key, course):
-            if certs_api.get_active_web_certificate(course) is not None:
-                context.update({
-                    'show_cert_web_view': True,
-                    'cert_web_view_url': certs_api.get_certificate_url(course_id=course_key, uuid=cert_status['uuid']),
-                })
-            else:
-                context.update({
-                    'is_downloadable': False,
-                    'is_generating': True,
-                    'download_url': None
-                })
-
     with outer_atomic():
         response = render_to_response('courseware/progress.html', context)
 
     return response
 
 
+def _get_cert_data(student, course, course_key, is_active, enrollment_mode):
+    """Returns students course certificate related data.
+
+    Arguments:
+        student (User): Student for whom certificate to retrieve.
+        course (Course): Course object for which certificate data to retrieve.
+        course_key (CourseKey): Course identifier for course.
+        is_active (Bool): Boolean value to check if course is active.
+        enrollment_mode (String): Course mode in which student is enrolled.
+
+    Returns:
+        returns dict if course certificate is available else None.
+    """
+    show_generate_cert_btn = (
+        is_active and CourseMode.is_eligible_for_certificate(enrollment_mode)
+        and certs_api.cert_generation_enabled(course_key)
+    )
+
+    show_cert_web_view = False
+    download_url = None
+    cert_web_view_url = None
+
+    if show_generate_cert_btn:
```

Comment:
I'd encourage two things in this block:

1. Whenever the code has enough information to return to the user, return immediately, rather than setting a local variable and waiting for the `if` block to terminate. That will make the control flow clearer, and make it less likely that something accidentally gets overridden later.
2. Once you've done 1, then rather than using a style with nested `if ... else ...` blocks, you can `return` from inside the `if` clause, and then leave off the `else` clause. That will reduce the level of nesting in the function, which will likely serve to make it more readable.
3. To optimize 2, aim for the `if` clauses to be the exceptional states that give you specific information and return values, so that you know you can immediately return from them. For example, rather than:

    if show_generate_cert_btn:
        ...
    elif enrollment_status == CourseMode.AUDIT:
        ...
    else:
        return None,

you could do:

    if enrollment_status == CourseMode.AUDIT:
        # return AUDIT message
    
    if not show_generate_cert_btn:
        return None

    ... all of the other logic ...